src/tools/miri/src/shims/foreign_items.rs

```diff
@@ -69,67 +78,102 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
             None =>
                 match link_name.as_str() {
                     "miri_start_panic" => {
+                        let args = this.copy_fn_args(args)?; // FIXME: Should `InPlace` arguments be reset to uninit?
+
                         // `check_shim` happens inside `handle_miri_start_panic`.
-                        this.handle_miri_start_panic(abi, link_name, args, unwind)?;
+                        this.handle_miri_start_panic(abi, link_name, &args, unwind)?;
                         return Ok(None);
                     }
                     // This matches calls to the foreign item `panic_impl`.
                     // The implementation is provided by the function with the `#[panic_handler]` attribute.
                     "panic_impl" => {
                         // We don't use `check_shim` here because we are just forwarding to the lang
-                        // item. Argument count checking will be performed when the returned `Body` is
-                        // called.
+                        // item. Argument count checking will be performed in `eval_fn_call`.
                         this.check_abi_and_shim_symbol_clash(abi, Abi::Rust, link_name)?;
                         let panic_impl_id = tcx.lang_items().panic_impl().unwrap();
                         let panic_impl_instance = ty::Instance::mono(tcx, panic_impl_id);
-                        return Ok(Some((
-                            this.load_mir(panic_impl_instance.def, None)?,
-                            panic_impl_instance,
-                        )));
+
+                        this.eval_fn_call(
+                            FnVal::Instance(panic_impl_instance),
+                            (abi, fn_abi),
+                            args,
+                            false,
+                            dest,
+                            ret,
+                            unwind,
+                        )?;
+
+                        return Ok(Some(()));
                     }
                     #[rustfmt::skip]
-                    | "exit"
+                    "exit"
                     | "ExitProcess"
                     => {
                         let exp_abi = if link_name.as_str() == "exit" {
                             Abi::C { unwind: false }
                         } else {
                             Abi::System { unwind: false }
                         };
-                        let [code] = this.check_shim(abi, exp_abi, link_name, args)?;
+                        let args = this.copy_fn_args(args)?; // FIXME: Should `InPlace` arguments be reset to uninit?
```

Comment:
Why is this duplicated so often now?

Reply:
Because there is a code path which needs `&[FnArg]`, but most need a `&[OpTy]`.

Reply:
There's now more of these asserts than before though, I think? So what changed?